### PR TITLE
feat(feature): handle feature update event from mercury

### PR DESCRIPTION
**Pull request by aimeex3**
handle event:featureToggle_update from mercury

# Pull Request Template

## Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Test Coverage

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] hooked up web-client with local sdk build, changed global notifications settings, tested that the mercury event gets triggered, and updates the feature collection

**Test Configuration**:
* SDK Version
* Node/Browser Version
* NPM Version

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/node_modules/@webex/internal-plugin-feature/src/feature.js [hunk 1]
```diff
@@ -37,6 +37,22 @@ const Feature = WebexPlugin.extend({
     return Promise.resolve(feature.value);
   },
 
+  /**
+   * Issues request to server to set a value for a feature toggle.
```
Comment (andrewholsted) [inline packages/node_modules/@webex/internal-plugin-feature/src/feature.js]: is this jsdoc correct?

File: packages/node_modules/@webex/internal-plugin-feature/src/feature.js [hunk 2]
```diff
@@ -37,6 +37,22 @@ const Feature = WebexPlugin.extend({
...
+   * @param {Object} envelope
+   * @returns {Promise} Refreshes the local device and resolves with the features endpoint's response.
+   */
+  handleFeatureUpdate(envelope) {
+    if (envelope && envelope.data) {
+      const feature = cloneDeep(envelope.data.featureToggle);
```
Comment (andrewholsted) [inline packages/node_modules/@webex/internal-plugin-feature/src/feature.js]: why do we need to clone here? Do we mutate the object somewhere else?

File: packages/node_modules/@webex/internal-plugin-feature/src/feature.js [hunk 3]
```diff
@@ -93,6 +109,10 @@ const Feature = WebexPlugin.extend({
   initialize(...args) {
     Reflect.apply(WebexPlugin.prototype.initialize, this, args);
 
+    this.webex.once('ready', () => {
```
Comment (andrewholsted) [inline packages/node_modules/@webex/internal-plugin-feature/src/feature.js]: Is it possible that the `ready` event has fired before this plugin has initialized? I don't think so, but wanted to check.

File: packages/node_modules/@webex/internal-plugin-feature/src/feature.js (hunk 1 above)
Comment (aimeex3) [reply packages/node_modules/@webex/internal-plugin-feature/src/feature.js]: oops forgot to update. thanks for catching

File: packages/node_modules/@webex/internal-plugin-feature/src/feature.js (hunk 2 above)
Comment (aimeex3) [reply packages/node_modules/@webex/internal-plugin-feature/src/feature.js]: yeah when it gets added to the feature collection, it parses the object and mutates it. I see 2 events with the same id come through here. Not sure why there's 2 and the second one is the one that gets mutated

File: packages/node_modules/@webex/internal-plugin-feature/src/feature.js (hunk 3 above)
Comment (aimeex3) [reply packages/node_modules/@webex/internal-plugin-feature/src/feature.js]: not sure, i thought ready meant that the plugins have been initialized. is there another event to check explicitly that this plugin is initialized? @mattnorris @taymoork2?

File: packages/node_modules/@webex/internal-plugin-feature/src/feature.js (hunk 2 above)
Comment (aimeex3) [reply packages/node_modules/@webex/internal-plugin-feature/src/feature.js]: i see where it's getting emitted twice: https://github.com/webex/webex-js-sdk/blob/master/packages/node_modules/@webex/internal-plugin-mercury/src/mercury.js#L442-L443

I think maybe server usually sends one or the other, but this looks like it sends both so it's the same object

Comment (aimeex3) [reply packages/node_modules/@webex/internal-plugin-feature/src/feature.js]: should i add a check in mercury.js, if the `namespace` === `data.eventType`, only emit 1 event? @mattnorris @taymoork2 @adamweeks

Comment (adamweeks) [reply packages/node_modules/@webex/internal-plugin-feature/src/feature.js]: @aimeex3 I think that should work!

File: packages/node_modules/@webex/internal-plugin-feature/src/feature.js (hunk 3 above)
Comment (taymoork2) [reply packages/node_modules/@webex/internal-plugin-feature/src/feature.js]: `plugin-meetings` does use `this.webex.once('ready'...` internally
Unless matt has something to add, I'd ask @arun3528 why they used it internally

Comment (mattnorris) [reply packages/node_modules/@webex/internal-plugin-feature/src/feature.js]: `ready` fires after *all* the plugins have initialized. 

Is there a reason not to use `listenToAndRun` like below?

Comment (aimeex3) [reply packages/node_modules/@webex/internal-plugin-feature/src/feature.js]: hmm even if `this.webex.internal.mercury` was not defined yet, `listenToAndRun` would work? I think mercury wasn't defined yet so i wanted to wait for it to be initialized

**Comment by aimeex3** [comment]
merged

**Pull request closed without merging**